File: electron/main/tools/pathValidation.ts
import { normalize, resolve } from 'path';
import { realpath } from 'fs/promises';
import { homedir } from 'os';
import { promises as fs } from 'fs';

/**
 * Security configuration for file access
 */
const SECURITY_CONFIG = {
  // Allowed base directories for file operations
  allowedDirectories: [
    homedir(), // User's home directory
    process.cwd(), // Current working directory
  ],
  // Blocked paths (absolute paths that should never be accessible)
  blockedPaths: [
    'C:\\Windows',
    'C:\\Program Files',
    'C:\\Program Files (x86)',
    'C:\\ProgramData',
    '/etc',
    '/usr/bin',
    '/usr/sbin',
    '/bin',
    '/sbin',
    '/system',
  ],
};

/**
 * Validates if a path is safe to access
 * @param requestedPath The path requested by the user/agent
 * @returns The validated absolute path
 * @throws Error if path is not allowed
 */
export async function validatePath(requestedPath: string): Promise<string> {
  if (!requestedPath) {
    throw new Error('Path cannot be empty');
  }

  // Normalize and resolve to absolute path
  // If path is relative, resolve it relative to user's home directory
  let absolutePath = resolve(requestedPath);
  
  // If the path doesn't look absolute (e.g., "Downloads"), try resolving from home
  if (!requestedPath.includes(':') && !requestedPath.startsWith('/') && !requestedPath.startsWith('\\')) {
    absolutePath = resolve(homedir(), requestedPath);
  }
  
  // Check if path exists first before trying realpath
  try {
    await fs.access(absolutePath);
  } catch {
    // If it's a write operation, the file might not exist yet, but we need to check the parent.
    // However, this function is general.
    // We'll throw a specific error.
    throw new Error(`File System Error: The path '${absolutePath}' does not exist.`);
  }
  
  // Resolve symlinks to prevent symlink traversal attacks
  const realPath = await realpath(absolutePath);
  
  // Check against blocked paths
  for (const blocked of SECURITY_CONFIG.blockedPaths) {
    if (realPath.toLowerCase().startsWith(blocked.toLowerCase())) {
      throw new Error(`Security Violation: Access to system directory '${realPath}' is strictly prohibited.`);
    }
  }
  
  // Check if path is within allowed directories
  let isAllowed = false;
  for (const allowedDir of SECURITY_CONFIG.allowedDirectories) {
    const resolvedAllowed = await realpath(allowedDir).catch(() => allowedDir);
    if (realPath.toLowerCase().startsWith(resolvedAllowed.toLowerCase())) {
      isAllowed = true;
      break;
    }
  }
  
  if (!isAllowed) {
    throw new Error(
      `Security Violation: Access to '${realPath}' is denied. Operations are restricted to the User Home Directory and the Project Directory.`
    );
  }
  
  return realPath;
}

/**
 * Validates a path for file reading operations
 */
export async function validateReadPath(path: string): Promise<string> {
  const validatedPath = await validatePath(path);
  
  // Check if file exists and is readable
  try {
    const stats = await fs.stat(validatedPath);
    if (stats.isDirectory()) {
      throw new Error(`Operation Failed: The path '${validatedPath}' is a directory, not a file.`);
    }
  } catch (error) {
    if (error instanceof Error && error.message.includes('no such file')) {
      throw new Error(`File Not Found: The file '${validatedPath}' could not be found.`);
    }
    throw error;
  }
  
  return validatedPath;
}

/**
 * Validates a path for directory listing operations
 */
export async function validateDirectoryPath(path: string): Promise<string> {
  const validatedPath = await validatePath(path);
  
  // Check if directory exists and is accessible
  try {
    const stats = await fs.stat(validatedPath);
    if (!stats.isDirectory()) {
      throw new Error(`Operation Failed: The path '${validatedPath}' is not a directory.`);
    }
  } catch (error) {
    if (error instanceof Error && error.message.includes('no such file')) {
      throw new Error(`Directory Not Found: The directory '${validatedPath}' could not be found.`);
    }
    throw error;
  }
  
  return validatedPath;
}

/**
 * Validates a path for file writing operations
 */
export async function validateWritePath(path: string): Promise<string> {
  if (!path) {
    throw new Error('Path cannot be empty');
  }

  // Resolve to an absolute path using the same rules as validatePath,
  // but do not require the file itself to exist.
  let absolutePath = resolve(path);
  if (!path.includes(':') && !path.startsWith('/') && !path.startsWith('\\')) {
    absolutePath = resolve(homedir(), path);
  }

  // Validate the parent directory (must exist, be a directory, and be within allowed roots).
  const parentDir = resolve(absolutePath, '..');
  
  try {
    await validateDirectoryPath(parentDir);
  } catch (error) {
      if (error instanceof Error && error.message.includes('Directory Not Found')) {
          throw new Error(`Operation Failed: The parent directory '${parentDir}' does not exist.`);
      }
      throw error;
  }

  // If the target file already exists, resolve its real path and fully validate it
  // to defend against symlink attacks.
  try {
    const fileRealPath = await realpath(absolutePath);
    return await validatePath(fileRealPath);
  } catch (error: any) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
      // New file: parent has been validated, so this path is safe to use.
      return absolutePath;
    }
    throw error;
  }
}

File: backend/tools/path_validation.py
import re
from pathlib import Path, PureWindowsPath
from urllib.parse import unquote

SECURITY_CONFIG = {
    "allowed_directories": [
        Path.home(),
        Path.cwd(),
    ],
    "blocked_paths": [
        "C:\\Windows",
        "C:\\Program Files",
        "C:\\Program Files (x86)",
        "C:\\ProgramData",
        "/etc",
        "/usr/bin",
        "/usr/sbin",
        "/bin",
        "/sbin",
        "/system",
    ],
}


def _is_within(path: Path, parent: Path) -> bool:
    """
    True if `path` is `parent` or a descendant of `parent`.
    Uses `Path.is_relative_to` to avoid string-prefix bypasses
    (e.g. `/home/snattack` "startswith" `/home/sn`).
    """
    try:
        return path == parent or path.is_relative_to(parent)
    except (ValueError, OSError):
        return False


def _resolved_allowed_directories() -> list[Path]:
    resolved: list[Path] = []
    for allowed in SECURITY_CONFIG["allowed_directories"]:
        try:
            resolved.append(allowed.resolve())
        except (OSError, RuntimeError):
            continue
    return resolved


def _is_blocked(path: Path) -> bool:
    """
    True if `path` falls under any configured blocked root.
    Blocked entries include Windows paths that are meaningless on Unix
    (and vice-versa); those simply won't match anything on the running platform,
    which is the intended behavior.
    """
    for blocked_str in SECURITY_CONFIG["blocked_paths"]:
        blocked = Path(blocked_str)
        if _is_within(path, blocked):
            return True
    return False


def _decode_requested_path(requested_path: str) -> str:
    if not isinstance(requested_path, str) or requested_path == "":
        raise ValueError("Path cannot be empty")
    if "\x00" in requested_path:
        raise ValueError("Path contains invalid characters")
    if re.search(r"%(?![0-9a-fA-F]{2})", requested_path):
        raise ValueError("Path contains invalid URL encoding")
    decoded_path = unquote(requested_path)
    if "\x00" in decoded_path:
        raise ValueError("Path contains invalid characters")
    return decoded_path


def _has_windows_anchor(path: str) -> bool:
    windows_path = PureWindowsPath(path)
    return bool(windows_path.drive or path.startswith("\\\\"))


def _reject_escape_components(path: Path) -> None:
    if any(part == ".." for part in path.parts):
        raise PermissionError("Security Violation: Path traversal is not permitted.")


def _resolve_requested_path(requested_path: str) -> Path:
    decoded_path = _decode_requested_path(requested_path)

    path = Path(decoded_path)
    _reject_escape_components(path)

    if not path.is_absolute():
        path = Path.home() / path

    resolved_path = path.resolve()

    # Reject Windows absolute/UNC paths only if they're not in allowed directories
    if _has_windows_anchor(str(resolved_path)):
        for allowed_dir in _resolved_allowed_directories():
            if _is_within(resolved_path, allowed_dir):
                return resolved_path
        raise PermissionError("Security Violation: Windows absolute paths are not permitted.")

    return resolved_path


def _ensure_allowed_path(real_path: Path) -> Path:
    if _is_blocked(real_path):
        raise PermissionError("Security Violation: Access to system directories is prohibited.")

    for allowed_dir in _resolved_allowed_directories():
        if _is_within(real_path, allowed_dir):
            return real_path

    raise PermissionError(
        "Security Violation: Access is restricted to the configured workspace."
    )


def validate_path(requested_path: str) -> Path:
    """Validate a readable existing path and return its resolved absolute path."""
    real_path = _resolve_requested_path(requested_path)
    real_path = _ensure_allowed_path(real_path)

    if not real_path.exists():
        raise FileNotFoundError("File System Error: The requested path does not exist.")

    return real_path


def validate_read_path(path: str) -> Path:
    """Validates a path for file reading operations."""
    validated_path = validate_path(path)

    if validated_path.is_dir():
        raise ValueError("Operation Failed: The requested path is a directory, not a file.")

    return validated_path


def validate_directory_path(path: str) -> Path:
    """Validates a path for directory listing operations."""
    validated_path = validate_path(path)

    if not validated_path.is_dir():
        raise ValueError("Operation Failed: The requested path is not a directory.")

    return validated_path


def validate_write_path(path: str) -> Path:
    """
    Validates a path for file writing operations.

    For existing targets, checks existence and enforces both the block-list
    and the allow-list directly. For new targets, validates
    the parent directory (which enforces allow-list transitively) and then
    re-checks the resolved path against both lists so the allow-list is never skipped.
    """
    resolved_path = _resolve_requested_path(path)

    if resolved_path.exists():
        _ensure_allowed_path(resolved_path)
        if resolved_path.is_dir():
            raise ValueError("Operation Failed: The requested path is a directory, not a file.")
        return resolved_path

    parent_dir = resolved_path.parent
    _ensure_allowed_path(parent_dir)
    if not parent_dir.is_dir():
        raise ValueError("Operation Failed: The parent directory does not exist.")

    return _ensure_allowed_path(resolved_path)
